crates/biome_js_analyze/src/lint/nursery/no_useless_undefined.rs

```diff
@@ -0,0 +1,303 @@
+    ///
+    /// ```js,expect_diagnostic
+    /// function* foo() {
+    ///   yield undefined;
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo(bar = undefined) {}
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo({bar = undefined}) {}
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo(undefined);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// let foo;
+    /// const {foo} = bar;
+    /// function foo() {
+    ///   return;
+    /// }
+    /// function* foo() {
+    ///   yield;
+    /// }
+    /// function foo(bar) {}
+    /// function foo({bar}) {}
+    /// foo();
+    /// ```
+    ///
+    pub NoUselessUndefined {
+        version: "next",
+        name: "noUselessUndefined",
+        language: "js",
+        fix_kind: FixKind::Safe,
+        sources: &[RuleSource::EslintUnicorn("no-useless-undefined")],
+        recommended: false,
+    }
+}
+
+declare_node_union! {
+    pub AnyUndefinedNode = JsVariableStatement
+        | JsObjectBindingPatternShorthandProperty
+        | JsYieldArgument
+        | JsReturnStatement
+        | JsArrayBindingPatternElement
+        | JsArrowFunctionExpression
+        | JsFormalParameter
+}
+
+fn find_undefined_range(expr: Option<&AnyJsExpression>) -> Option<TextRange> {
+    let ident = expr?.as_js_reference_identifier()?;
+    if ident.is_undefined() {
+        Some(ident.range())
+    } else {
+        None
+    }
+}
+
+pub struct RuleState {
+    binding_text: Option<Box<str>>,
+    diagnostic_range: TextRange,
+}
+
+impl Rule for NoUselessUndefined {
+    type Query = Ast<AnyUndefinedNode>;
+    type State = RuleState;
+    type Signals = Box<[Self::State]>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let mut signals = vec![];
+
+        match node {
+            // let foo = undefined;
+            AnyUndefinedNode::JsVariableStatement(statement) => {
+                let Ok(node) = statement.declaration() else {
+                    return signals.into_boxed_slice();
+                };
+                let let_or_var = node.is_let() || node.is_var();
+                if !let_or_var {
+                    return signals.into_boxed_slice();
+                }
+
+                for declarator in node.declarators() {
+                    let Ok(decl) = declarator else { continue };
+                    let Some(initializer) = decl.initializer() else {
+                        continue;
+                    };
+                    let expr = initializer.expression().ok();
+                    if let Some(undefined_range) = find_undefined_range(expr.as_ref()) {
+                        let Ok(binding_text) = decl.id() else {
+                            continue;
+                        };
+                        signals.push(RuleState {
+                            binding_text: Some(binding_text.text().into()),
```

Comment:
If we opt for `TokenText` (See previous comment):
```suggestion
                            binding_text: Some(binding_text.token_text_trimmed()),
```

Reply:
Thank you, I've updated it at [4b43a02](https://github.com/biomejs/biome/pull/4100/commits/4b43a02bf62241f4eca1fb566da7d619354df3b4)